<sdf version='1.4'>
  <world name='default'>
    <!-- A global light source -->
    <include>
      <uri>model://sun</uri>
    </include>

    <!-- A ground plane -->
    <include>
      <uri>model://ground_plane</uri>
    </include>
    
    <physics type="ode">
      <real_time_update_rate>1000.0</real_time_update_rate>
      <max_step_size>0.001</max_step_size>
      <real_time_factor>1</real_time_factor>
      <ode>
        <solver>
          <type>quick</type>
          <iters>150</iters>
          <precon_iters>0</precon_iters>
          <sor>1.400000</sor>
          <use_dynamic_moi_rescaling>1</use_dynamic_moi_rescaling>
        </solver>
        <constraints>
          <cfm>0.00001</cfm>
          <erp>0.2</erp>
          <contact_max_correcting_vel>2000.000000</contact_max_correcting_vel>
          <contact_surface_layer>0.01000</contact_surface_layer>
        </constraints>
      </ode>
    </physics>

    <!-- Load world -->
    <road name="Lanes_4">
      <width>15</width>
      <point>-100 0 0.005</point>
      <point>100 0 0.005</point>
      <material>
        <script>
          <uri>file://media/materials/scripts/gazebo.material</uri>
          <name>Gazebo/Lanes_4</name>
        </script>
      </material>
    </road>

    <scene>
      <ambient>0.4 0.4 0.4 1</ambient>
      <background>0.7 0.7 0.7 1</background>
      <shadows>true</shadows>
    </scene>

    <gui fullscreen='0'>
      <camera name='user_camera'>
        <pose>0.8 0.0 30.0 0 1.5708 0</pose>
        <view_controller>orbit</view_controller>
      </camera>
    </gui>

    <population name="left_construction_barrel_population">
      <model name="left_construction_barrel">
        <include>
          <static>true</static>
          <uri>model://construction_barrel</uri>
        </include>
      </model>
      <pose>20 8 0 0 0 0</pose>
      <distribution>
        <type>grid</type>
        <rows>1</rows>
        <cols>10</cols>
        <step>3 1 0</step>
      </distribution>
    </population>

    <population name="right_construction_barrel_population">
      <model name="right_construction_barrel">
        <include>
          <static>true</static>
          <uri>model://construction_barrel</uri>
        </include>
      </model>
      <pose>20 -8 0 0 0 0</pose>
      <distribution>
        <type>grid</type>
        <rows>1</rows>
        <cols>10</cols>
        <step>3 1 0</step>
      </distribution>
    </population>    

    <model name="driving_terrain">
      <include>
        <static>true</static>
        <uri>model://vrc_driving_terrain</uri>
      </include>
    </model>

    <!-- ************************************************************************************************ -->

    <model name="opponent_car_hatchback_blue">
      <pose> 9 1.667 0.01 0 0 0</pose>
      <include>
            <static>false</static>
        <uri>model://hatchback_blue</uri>
      </include>

      <plugin name="object_controller" filename="libgazebo_ros_planar_move.so">
        <commandTopic>/opponent_car_hatchback_blue/cmd_vel</commandTopic>
        <odometryTopic>/opponent_car_hatchback_blue/odom</odometryTopic>
        <odometryFrame>/opponent_car_hatchback_blue/odom</odometryFrame>
        <odometryRate>10.0</odometryRate>
        <robotBaseFrame>base_footprint</robotBaseFrame>
      </plugin>

    </model>

    <model name="opponent_car_hatchback_red">
      <pose> 9 -5 0.01 0 0 0</pose>
      <include>
            <static>false</static>
        <uri>model://hatchback_red</uri>
      </include>

      <plugin name="object_controller" filename="libgazebo_ros_planar_move.so">
        <commandTopic>/opponent_car_hatchback_red/cmd_vel</commandTopic>
        <odometryTopic>/opponent_car_hatchback_red/odom</odometryTopic>
        <odometryFrame>/opponent_car_hatchback_red/odom</odometryFrame>
        <odometryRate>10.0</odometryRate>
        <robotBaseFrame>base_footprint</robotBaseFrame>
      </plugin>

    </model>

    <model name="opponent_car_hatchback_grey">
      <pose> 9 -1.667 0.01 0 0 0</pose>
      <include>
            <static>false</static>
        <uri>model://hatchback</uri>
      </include>

      <plugin name="object_controller" filename="libgazebo_ros_planar_move.so">
        <commandTopic>/opponent_car_hatchback_grey/cmd_vel</commandTopic>
        <odometryTopic>/opponent_car_hatchback_grey/odom</odometryTopic>
        <odometryFrame>/opponent_car_hatchback_grey/odom</odometryFrame>
        <odometryRate>10.0</odometryRate>
        <robotBaseFrame>base_footprint</robotBaseFrame>
      </plugin>

    </model>

    <model name="ego">
      <pose> 0 -1.667 0.01 0 0 0</pose>
      <include>
            <static>false</static>
        <uri>model://hatchback_ego</uri>
      </include>

      <plugin name="object_controller" filename="libgazebo_ros_planar_move.so">
        <commandTopic>/ego/cmd_vel</commandTopic>
        <odometryTopic>/ego/odom</odometryTopic>
        <odometryFrame>/ego/odom</odometryFrame>
        <odometryRate>20.0</odometryRate>
        <robotBaseFrame>base_footprint</robotBaseFrame>
      </plugin>

    </model>


  </world>
</sdf>
